src/llama.cpp

```diff
@@ -17592,11 +17593,10 @@ struct llama_context * llama_new_context_with_model(
         } else {
             // LLAMA_SPLIT_LAYER requires a backend for each GPU
             for (int i = 0; i < ggml_backend_sycl_get_device_count(); ++i) {
-                ggml_backend_t backend = ggml_backend_sycl_init(i);
+                int id = ggml_backend_sycl_get_device_id(i);
```

Comment:
why again there is difference between ```i``` and ```id```?

Reply:
There is std::vector<int> to save the ID of avaible devices.

`i` is the index to access vector of available devices.
`id` is the ID of SYCL device saved in vector.

To go through the available devices, there are two methods:
1. for (int i=0;i< devices.size();i++) {
    int id = devices[i]; 
   }

2. for(auto &id: devices) {
   ...
   }

I choose item 1.

Reply:
SYCL backend save the original device IDs, like 0, 1, 2, 3, 4, 5. They are saved in static instance of DPCT.
When user choose 0,2 by environment variable SYCL_VISIBLE_DEVICE, dev_gpu_mgr will save the avaible devices as [0, 2].

The original device IDs are not changed, as they are saved in static instance of DPCT.
So, user still can see all devices.
But SYCL backend will operate device 0, 2 only.

To implement your suggestion, DPCT static instance need to be refreshed. 
After do it, ID 2 will be unavailable, because the available device ID will become 0, 1. 
It will be in a loop.
I keep the device IDs in DPCT are not changed as base.
So that the ID 0-5 are always be unique ID of SYCL devices.




When use ONEAPI_DEVICE_SELECTOR, DPCT and SYCL backend will see 0,2 only.
Because DPCT will be impacted by it too.

Reply:
then remove the so-called ```GGML_SYCL_VISIBLE_DEVICES```. Counterpart of ```CUDA_VISBLE_DEVICE``` is ```ONEAPI_DEVICE_SELECTOR```, don't need to define another environmental variable to confuse the users.

Reply:
ONEAPI_DEVICE_SELECTOR grammar is a little complex.
CUDA_VISBLE_DEVICE is friedly to normal user.
If only support ONEAPI_DEVICE_SELECTOR , there will be more usage issues about it.
CUDA_VISBLE_DEVICE style avoid to make mistake in usage.

Current implementation is refered to VULKAN backend too.
It's saved the available devices by std::vector<int> too.

Reply:
I don't know why ```ONEAPI_DEVICE_SELECTOR grammar is a little complex``` which is quite straightforward. Users must ensure they know the grammar well if they really want to set up multiply device. Otherwise they can use the default settings.

>Current implementation is refered to VULKAN backend.
>It's saved the available devices by std::vector too.

There is already a device manager under DPCT, why always add another device manager again? If you want another behaviour, update dev_mgr directly instead of adding one more device list to ggml-sycl.cpp which is quite complex already and we are working to simplify it.

Reply:
SYCL_VISIBLE_DEVICES improve the OOB, & it has same format of CUDA/VULKAN_VISIBLE_DEVICES.
I think the environment variable will be replaced by other method in the future.
Keep with same style will reduce the adaption work in the future.

The "another device manager" is a class to manage the choosen devices. Don't care the name, it could be renamed to another, like "chosen_devices()" or others. It has nothing with the DPCT::dev_mgr.  I just refer it's name in naming. :)

DPCT::dev_mgr is already complex. Define new class is easier than update existed complex class.

Reply:
> DPCT::dev_mgr is already complex. Define new class is easier than update existed complex class.

Going around problems rather than solving them creates more technical debt